src/sparsezoo/utils/data.py

```diff
@@ -28,10 +28,98 @@
 
 __all__ = ["Dataset", "RandomDataset", "DataLoader"]
 
-
 _LOGGER = logging.getLogger(__name__)
 
 
+#    A utility class to load data in batches for fixed number of iterations
+
+
+class _BatchLoader:
+    __slots__ = [
+        "_data",
+        "_batch_size",
+        "_iterations",
+        "_batch_buffer",
+        "_batch_template",
+        "_batches_created",
+    ]
+
+    def __init__(
+        self,
+        data: Iterable[Union[numpy.ndarray, List[numpy.ndarray]]],
+        batch_size: int,
+        iterations: int,
+    ):
+        self._data = data
+        _single_input = type(self._data[0]) is numpy.ndarray
+        if _single_input:
+            self._data = [self._data]
+        self._batch_size = batch_size
+        self._iterations = iterations
+        if batch_size < 0 or iterations < 0:
+            raise ValueError(
+                f"Both batch size and number of iterations should be non-negative, "
+                f"supplied values (batch_size, iterations):{(batch_size, iterations)}"
+            )
+
+        self._batch_buffer = []
+        self._batch_template = self._init_batch_template()
+        self._batches_created = 0
+
+    def __iter__(self) -> Generator[List[numpy.ndarray], None, None]:
+        yield from self._multi_input_batch_generator()
+
+    @property
+    def _buffer_is_full(self) -> bool:
+        return len(self._batch_buffer) == self._batch_size
+
+    @property
+    def _all_batches_loaded(self) -> bool:
+        return self._batches_created >= self._iterations
+
+    def _multi_input_batch_generator(
+        self,
+    ) -> Generator[List[numpy.ndarray], None, None]:
+        # A generator for with each element of the form
+        # [[(batch_size, features_a), (batch_size, features_b), ...]]
+        while not self._all_batches_loaded:
+            yield from self._batch_generator(source=self._data)
+
+    def _batch_generator(self, source) -> Generator[List[numpy.ndarray], None, None]:
+        # batches from source
+        for sample in source:
+            self._batch_buffer.append(sample)
+            if self._buffer_is_full:
+                _batch = self._make_batch()
+                yield _batch
```

Comment:
nit: can we add in functionality to unwrap the list into a single value if originally there were single inputs? (i.e. if `self._data` is a `List[numpy.ndarray]` then unwrap the single list, but if it happens to be a `List[List[numpy.ndarray]]` with the inner list size 1, then don't unwrap. This can be checked in the initializer for `_BatchLoader`

Reply:
Sure thang!